src/README.md

```diff
@@ -0,0 +1,753 @@
+
+`args.GetReturnValue()` is a placeholder for the return value of the function,
+and provides a `.Set()` method that can be called with a boolean, integer,
+floating-point number or a `Local<Value>` to set the return value.
+
+Node.js provides various helpers for building JS classes in C++ and/or attaching
+C++ functions to the exports of a built-in module:
+
+```c++
+void Initialize(Local<Object> target,
+                Local<Value> unused,
+                Local<Context> context,
+                void* priv) {
+  Environment* env = Environment::GetCurrent(context);
+
+  env->SetMethod(target, "getaddrinfo", GetAddrInfo);
+  env->SetMethod(target, "getnameinfo", GetNameInfo);
+
+  // 'SetMethodNoSideEffect' means that debuggers can safely execute this
+  // function for e.g. previews.
+  env->SetMethodNoSideEffect(target, "canonicalizeIP", CanonicalizeIP);
+
+  // ... more code ...
+
+  // Building the `ChannelWrap` class for JS:
+  Local<FunctionTemplate> channel_wrap =
+      env->NewFunctionTemplate(ChannelWrap::New);
+  // Allow for 1 internal field, see `BaseObject` for details on this:
+  channel_wrap->InstanceTemplate()->SetInternalFieldCount(1);
+  channel_wrap->Inherit(AsyncWrap::GetConstructorTemplate(env));
+
+  // Set various methods on the class (i.e. on the prototype):
+  env->SetProtoMethod(channel_wrap, "queryAny", Query<QueryAnyWrap>);
+  env->SetProtoMethod(channel_wrap, "queryA", Query<QueryAWrap>);
+  // ...
+  env->SetProtoMethod(channel_wrap, "querySoa", Query<QuerySoaWrap>);
+  env->SetProtoMethod(channel_wrap, "getHostByAddr", Query<GetHostByAddrWrap>);
+
+  env->SetProtoMethodNoSideEffect(channel_wrap, "getServers", GetServers);
+
+  Local<String> channel_wrap_string =
+      FIXED_ONE_BYTE_STRING(env->isolate(), "ChannelWrap");
+  channel_wrap->SetClassName(channel_wrap_string);
+  target->Set(env->context(), channel_wrap_string,
+              channel_wrap->GetFunction(context).ToLocalChecked()).Check();
+}
+
+// Run the `Initialize` function when loading this module through
+// `internalBinding('cares_wrap')` in Node.js’s built-in JavaScript code:
+NODE_MODULE_CONTEXT_AWARE_INTERNAL(cares_wrap, Initialize)
+```
+
+<a id="exception-handling"></a>
+### Exception handling
+
+The V8 engine provides multiple features to work with JavaScript exceptions,
+as C++ exceptions are disabled inside of Node.js:
+
+#### Maybe types
+
+V8 provides the `v8::Maybe<T>` and `v8::MaybeLocal<T>` types, typically used
+as return values from API functions that can run JavaScript code and therefore
+can throw exceptions.
+
+Conceptually, the idea is that every `v8::Maybe<T>` is either empty (checked
+through `.IsNothing()`) or holds a value of type `T` (checked through
+`.IsJust()`). If the `Maybe` is empty, then a JavaScript exception is pending.
+A typical way of accessing the value is using the `.To()` function, which
+returns a boolean indicating success of the operation (i.e. the `Maybe` not
```

Comment:
Find `(i.e. the Maybe ...)` a bit superfluous.

Reply:
@devnexen I was worried an unfamiliar reader would think this is about success of the `.To()` call, not of the call that resulted in the `Maybe`.

Reply:
Ok that's fine, I just felt it broke a bit the fluidity of the reading.